test/unit/ngsiv2/expressions/expressionBasedTransformations-test.js

```diff
@@ -0,0 +1,612 @@
+                    },
+                    {
+                        object_id: 'a',
+                        name: 'alive',
+                        type: 'Null',
+                        expression: '${@alive *  20}'
+                    },
+                    {
+                        object_id: 'u',
+                        name: 'updated',
+                        type: 'Boolean',
+                        expression: '${@updated *  20}'
+                    },
+                ]
+            },
+            'WeatherStationMultiple': {
+                commands: [],
+                type: 'WeatherStation',
+                lazy: [],
+                active: [
+                    {
+                        object_id: 'p',
+                        name: 'pressure25',
+                        type: 'Integer',
+                        expression: '${@pressure * 20}'
+                    },
+                    {
+                        object_id: 'h',
+                        name: 'humidity12',
+                        type: 'Percentage'
+                    },
+                    {
+                        name: 'weather',
+                        type: 'Summary',
+                        expression: 'Humidity ${@humidity12 / 2} and pressure ${@pressure25 * 20}'
+                    },
+                    {
+                        object_id: 'a',
+                        name: 'alive',
+                        type: 'Null',
+                        expression: '${trim(@spaces)}'
+                    },
+                    {
+                        object_id: 'u',
+                        name: 'updated',
+                        type: 'Boolean',
+                        expression: '${trim(@spaces)}'
+                    },
+                ]
+            }
+        },
+        service: 'smartGondor',
+        subservice: 'gardens',
+        providerUrl: 'http://smartGondor.com',
+        deviceRegistrationDuration: 'P1M',
+        throttling: 'PT5S'
+    };
+
+describe('Expression-based transformations plugin', function() {
+    beforeEach(function(done) {
+        logger.setLevel('FATAL');
+
+        iotAgentLib.activate(iotAgentConfig, function() {
+            iotAgentLib.clearAll(function() {
+                iotAgentLib.addUpdateMiddleware(iotAgentLib.dataPlugins.attributeAlias.update);
+                iotAgentLib.addQueryMiddleware(iotAgentLib.dataPlugins.attributeAlias.query);
+                iotAgentLib.addUpdateMiddleware(iotAgentLib.dataPlugins.expressionTransformation.update);
+                done();
+            });
+        });
+    });
+
+    afterEach(function(done) {
+        iotAgentLib.clearAll(function() {
+            iotAgentLib.deactivate(done);
+        });
+    });
+
+    describe('When an update comes for attributes with expressions', function() {
+        var values = [
+            {
+                name: 'p',
+                type: 'Integer',
+                value: 52
```

Comment:
Just to check if I'm understanding correctly the "separtion of concerns" between IOTA code and IOTA Lib code...

When the IOTA receives something like this in its southbound interface:

```
p|52
```

It is the responsability of the agent to grab the measure, check that "52" and cast it to a number before pasing it to the library. The library will get a 52, as shown in the `values` used to feed the test.

Is my interpretation correct?

Reply:
Let's going to see the workflow with a real example: iotagent-ul with MQTT binding.
1. A device publishes an active attribute in a topic as you suggest: 
```
p|52
```

2. The iotagent-ul applies the following handle from [commonBindings.js](https://github.com/telefonicaid/iotagent-ul/blob/master/lib/commonBindings.js):
```
function singleMeasure(apiKey, attribute, device, message) {
    var values;
    values = [
        {
            name: attribute,
            type: guessType(attribute, device),
            value: message.toString()
        }
    ];

    iotAgentLib.update(device.name, device.type, '', values, device, function(error) {
        if (error) {
            config.getLogger().error(context,
                ' MEASURES-002: Couldn\'t send the updated values to the Context Broker due to an error: %s', error);
        } else {
            config.getLogger().debug(context, 'Single measure for device [%s] with apiKey [%s] successfully updated',
                device.id, apiKey);
        }
    });
}
```

3. As you can see this agent does not apply any kind of checking or cast regarding the input value of the measures and rely on the IOTA Lib code. It just tries to convert the soutbound data model to NGSI, e.g. in this case from UL to NGSI. From this point, we are exactly in the same point that we have with the unit tests of the IOTA Lib code.
4. If there is not a plugin activated, the IOTA Lib code just sends the attributes as they are received, so if an attribute with an integer value is received for NGSIv1 version, it would be sent to the CB despite not being supported.
5. If there the expression plugin is enabled, the input value is passed to the expression parser which tries to applied the expression using the value as input parameter.
6. The output value of the expression parser is used to replace the input one using replace instruction which causes a cast to String type.

Reply:
Thanks for the detailed workflow! I have a doubt regarding

> As you can see this agent does not apply any kind of checking or cast regarding the input value of the measures and rely on the IOTA Lib code.

However, the value seems to be casted to string in the IOTA code:

```
value: message.toString()
```

Thus, for the IOTAgent `p|52` means always "52" (string) and never 52 (number).

(Maybe by "agent does not apply any kind of checking or cast" you mean what I refer as "casted to string"  Just clarifying... :)

Is this correct?

Reply:
Upsssss!! You are completely right, I missed that specific and relevant instruction because I was just trying to answer quickly. As you point, iotagent-ul is doing the cast. Anyway, I guess that it is not mandatory because in NGSIv1 is possible to use numbers, bools... although when they are stored by the CB, they are transformed to Strings,

For instance, in the LWM2M (commonLwm2m.js)[https://github.com/telefonicaid/lightweightm2m-iotagent/blob/master/lib/services/lwm2mHandlers/commonLwm2m.js]:

```
function activeDataHandler(registeredDevice, name, type, value) {
    var attributes = [{
        name: name,
        type: type,
        value: value
    }];

    logger.debug(context, 'Handling data from device [%s]', registeredDevice.id);

    iotAgentLib.update(
        registeredDevice.name,
        registeredDevice.type,
        '',
        attributes,
        registeredDevice,
        function handleUpdateEntity(error) {
            if (error) {
                logger.error(context, 'Unknown error connecting with the Context Broker: ' + error);
            } else {
                logger.debug(context, 'Data handled successfully');
            }
        }
    );
}
```

Reply:
So it seems that at the end the way in which the IOTAgent processes the attribute values before passing the to the IOTA Lib is a key point.

This has a important "coloraly": in order to support NGSIv2 in IOTAs it is not enough to support only the natives types allowed by NGSIv2 in the IOTA Lib. The code of the different IOTAs should be reviewed, e.g. IOTA-UL shouldn't do unconditional cast to string. 

Maybe an issue about "NGSIv2 migration" should be opened in each one of the target IOTAs (IOTA-UL and IOTA-JSON at least).

Reply:
Another doubt...

> The output value of the expression parser is used to replace the input one using replace instruction which causes a cast to String type.

If I understand correctly, you mean that if an expression is used, no matter the input of such expression, the output is always cast to String.

However, looking to your test updateContextExpressionPlugin1.json:

```
{
  "pressure": {
    "value": 1040,
    "type": "Integer"
  }
}
```

This is the result of applying the following

```
{
   object_id: 'p',
   name: 'pressure',
   type: 'Integer',
   expression: '${@pressure * 20}'
}
```

to value 52.

Should I understand that is `type` which defines which output type to use? I mean, if I use:


```
{
   object_id: 'p',
   name: 'pressure',
   type: 'String',
   expression: '${@pressure * 20}'
}
```

I'd get:

```
{
  "pressure": {
    "value": "1040",
    "type": "Integer"
  }
}
```

Is that correct?

Reply:
In NGSIv1, no matter the input of such expression, the output always cast to String. But now, with the modifications that we have introduced to support JSON native types, you'r right `type` defines which output type to use.

To sum up:

* Input attribute:
```
{
  "pressure": {
    "value": 52,
    "type": "Integer"
  }
}
```

* Expression:

```
{
   object_id: 'p',
   name: 'pressure',
   type: 'Integer',
   expression: '${@pressure * 20}'
}
```

* Transformed attribute with previous version of IOTA Lib:
```
{
  "pressure": {
    "value": "1040",
    "type": "Integer"
  }
}
```

* Transformed attribute with current version of IOTA Lib (supporting NGSIv2)
```
{
  "pressure": {
    "value": 1040,
    "type": "Integer"
  }
}
```

Reply:
Regarding:

> This has a important "coloraly": in order to support NGSIv2 in IOTAs it is not enough to support only the natives types allowed by NGSIv2 in the IOTA Lib. The code of the different IOTAs using the library should be reviewed, e.g. IOTA-UL shouldn't do unconditional cast to string.

> Maybe an issue about "NGSIv2 migration" should be opened in each one of the target IOTAs (IOTA-UL and IOTA-JSON at least)

I fully agree with you, we should open these issues since modifications will be needed in all the IOTAs.

Reply:
Issues opened at:

https://github.com/telefonicaid/iotagent-json/issues/250
https://github.com/telefonicaid/iotagent-ul/issues/233
https://github.com/telefonicaid/lightweightm2m-iotagent/issues/104

Reply:
I'm looking in detail de documentation and I read (with regards to NGSIv1):

> Expressions can have two return types: String or Number. This return type must be configured for each attribute that is going to be converted. Default value type is String.
>
> Whenever a expression is executed without error, its result will be cast to the configured type. If the conversion fails (e.g.: if the expression is null or a String and is casted to Number), the measurement update will fail, and an error will be reported to the device.

The "its result will be cast to the configured type" contradicts "In NGSIv1, no matter the input of such expression, the output always cast to String". Is the documentation wrong? In that case, it should be fixed.

If the documentation is ok, it should clarify which type is used: the one defined for the attribute in the API or the one to which the expression belongs (I mean, if the expression belong to "string operations" the type is String and if the expression belongs to "numeric operations" the type is Number).

Reply:
That part of the original documentation is correct... but only for the parse function of (expressionParser.js)[https://github.com/telefonicaid/iotagent-node-lib/blob/master/lib/plugins/expressionParser.js]. You can check in [expression-test.js](https://github.com/telefonicaid/iotagent-node-lib/blob/master/test/unit/expressions/expression-test.js), that this function complies with

> Whenever a expression is executed without error, its result will be cast to the configured type. If the conversion fails (e.g.: if the expression is null or a String and is casted to Number), the measurement update will fail, and an error will be reported to the device.

But how this function is used by [expressionPlugin.js](https://github.com/telefonicaid/iotagent-node-lib/blob/master/lib/plugins/expressionPlugin.js):
1. In line 60, for the entity, the function processExpressionAttributes of (expressionParser.js)[https://github.com/telefonicaid/iotagent-node-lib/blob/master/lib/plugins/expressionParser.js] is called.
2. processExpressionAttributes (line 179) function calls to expressionApplier
3. expressionApplier (line 157) function calls to applyExpression
4. applyExpression (line 145) function. Calls to processExpression and uses replace instruction to update the original value. As you can see in https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/replace, the replace() method returns a new string. Please find below the code of this function:

```
function applyExpression(expression, context, typeInformation) {
    var expressionList = expression.match(/\$\{.*?\}/g) || [],
        substitutions = expressionList.map(processExpression(context)),
        expressionResult = expression;

    for (var i = 0; i < substitutions.length; i++) {
        expressionResult = expressionResult.replace(substitutions[i].original, substitutions[i].value);
    }

    return expressionResult;
}
```

5. processExpression (line 131) function. Calls to parse function BUT it uses as argument the type 'String'. Therefore, the results of the expression will be always cast to String.

Do you see my point?? Sorry, maybe I am not explaining well how this is implemented but it's a bit tricky....

Reply:
Looking to expression-test.js file, it seems that the parse() function allows to define the output type. There is a function parameter for that, e.g.:

```
expressionParser.parse(arithmeticExpr[0], scope, 'Number', function(error, result)
...
expressionParser.parse('"Pruebas" # "DeStrings"', scope, 'String', function(error, result)
```

Next, looking to the logic used by the expressionPlugin.js (the one you perfectly describe in the 5 steps) I understand that the "link" with parse() is in the following point (https://github.com/telefonicaid/iotagent-node-lib/blob/master/lib/plugins/expressionParser.js#L136, processExpression() function):

```
 result = parse(cleanedExpression, context, 'String');
```

So it seems that although we have a poweful function able to cast to Numbers and String we are capping ourselves to use only the String mode :)

If my interpretation is correct (please confirm... as you say this is tricky so let's confirm common understading in each step in the dicussion :) I'd suggest the following modifications:

1. Add a comment just above L136 in order to explain that the function is able to use String and Number but we are using only String. Suggestion (please freely edit the text):

```
// Note that parse() function allows both String way processing and Number way processing
// (have a look to  expression-test.js). However, here we only use one of the possibilities, i.e. String
```

2. Adjust https://github.com/dcalvoalonso/iotagent-node-lib/blob/master/doc/expressionLanguage.md#-types section given that it is confusing (or even wrong). Suggestion (please review if I'm right in the last paragraph):

> The way the `parse()` function works (at expressionParser.js) is as follows:
>
> * Expressions can have two return types: String or Number. This return type must be configured for each attribute that is going to be converted. Default value type is String.
> * Whenever a expression is executed without error, its result will be cast to the configured type. If the conversion fails (e.g.: if the expression is null or a String and is casted to Number), the measurement update will fail, and an error will be reported to the device.
> 
> However, the usage that the Transformation Expression does of that function is using always String type. That means that at the end the resulf of the expression will be always casted to String. However, in NGSIv2 that String result could be re-cast to the right type (i.e. the one defined for the attribute in the provision operation). Have a look to the [NGSIv2 section] for more information on this.

Reply:
Your analysis is correct so I guess that we have now a common understanding of this plugin! :)

I will update the code to include your suggestions. Thanks!!!

Reply:
Fixed in 45d40dd86730117e13f4e62b5dd8191e51b5bc6f